test/e2e/framework/metrics/metrics_grabber.go

```diff
@@ -102,37 +123,48 @@ func NewMetricsGrabber(c clientset.Interface, ec clientset.Interface, kubelets b
 			break
 		}
 	}
-	if kubeScheduler == "" {
-		scheduler = false
-		klog.Warningf("Can't find kube-scheduler pod. Grabbing metrics from kube-scheduler is disabled.")
-	}
-	if kubeControllerManager == "" {
-		controllers = false
-		klog.Warningf("Can't find kube-controller-manager pod. Grabbing metrics from kube-controller-manager is disabled.")
-	}
-	if snapshotControllerManager == "" {
-		snapshotController = false
-		klog.Warningf("Can't find snapshot-controller pod. Grabbing metrics from snapshot-controller is disabled.")
-	}
-	if ec == nil {
+	if clusterAutoscaler && ec == nil {
 		klog.Warningf("Did not receive an external client interface. Grabbing metrics from ClusterAutoscaler is disabled.")
 	}
 
 	return &Grabber{
 		client:                     c,
 		externalClient:             ec,
+		config:                     config,
 		grabFromAPIServer:          apiServer,
-		grabFromControllerManager:  controllers,
+		grabFromControllerManager:  checkPodDebugHandlers(c, controllers, "kube-controller-manager", kubeControllerManager),
 		grabFromKubelets:           kubelets,
-		grabFromScheduler:          scheduler,
+		grabFromScheduler:          checkPodDebugHandlers(c, scheduler, "kube-scheduler", kubeScheduler),
 		grabFromClusterAutoscaler:  clusterAutoscaler,
-		grabFromSnapshotController: snapshotController,
+		grabFromSnapshotController: checkPodDebugHandlers(c, snapshotController, "snapshot-controller", snapshotControllerManager),
 		kubeScheduler:              kubeScheduler,
 		kubeControllerManager:      kubeControllerManager,
 		snapshotController:         snapshotControllerManager,
 	}, nil
 }
 
+func checkPodDebugHandlers(c clientset.Interface, requested bool, component, podName string) bool {
```

Comment:
is not a bit weird to have `requested` here? if the caller knows he is not going to request info why he calls the function anyway?
I see that it simplifies the code in `NewMetricsGrabber` constructor but is it worth it?

Reply:
> I see that it simplifies the code in NewMetricsGrabber constructor but is it worth it?

This is an internal helper function. It's whole purpose is to make the constructor simpler and not have any repeated code there (like this if check), so I think it's worth it.

Reply:
> if the caller knows he is not going to request info why he calls the function anyway?

The caller is the constructor and it doesn't know whether metrics for a component got requested. It's repeating that if check in the constructor that gets avoided by moving it into a parameter.